utilities/src/main/java/ru/study21/jcsv/xxl/algorithms/ExternalKwayMerge.java

```diff
@@ -0,0 +1,341 @@
+                CachedNioBinaryWriter writer = new CachedNioBinaryWriter(binOutputChannel, writingCacheSize)
+        ) {
+            // go go go
+            doKwayMerge(
+                    reader,
+                    writer,
+                    regions.size()
+            );
+        }
+    }
+
+    private void doublePassKwayMerge(
+            SeekableByteChannel binInputChannel,
+            SeekableByteChannel binOutputChannel,
+            long approxMemLimit,
+            long presortedLen,
+            int k1, // k for first pass
+            int writingCacheSize
+    ) throws IOException {
+        approxMemLimit -= writingCacheSize;
+        if (approxMemLimit < 0) {
+            throw new IllegalArgumentException("not enough memory provided");
+        }
+        Path tempFile = FileManager.createTempDirectory("singlePassKwayMerge").createTempFile("temp");
+
+        // first pass writes to temporary file
+        try (SeekableByteChannel tempChannel = Files.newByteChannel(tempFile, StandardOpenOption.WRITE)) {
+            long inputSize = binInputChannel.size();
+            int readCacheSize = (int) (approxMemLimit / k1);
+
+            for (long firstPassStart = 0; firstPassStart < inputSize; firstPassStart += k1 * presortedLen) {
+                // setup reader
+                List<MultiregionCachedNioBinaryReader.Region> regions = new ArrayList<>(k1);
+                long partEnd = Math.min(inputSize, firstPassStart + k1 * presortedLen);
+                for (long r = firstPassStart; r <= partEnd - presortedLen; r += presortedLen) {
+                    regions.add(new MultiregionCachedNioBinaryReader.Region(r, presortedLen));
+                }
+                if (partEnd % presortedLen != 0) {
+                    long lastLen = partEnd % presortedLen;
+                    regions.add(new MultiregionCachedNioBinaryReader.Region(partEnd - lastLen, lastLen));
+                }
+
+                try (
+                        MultiregionCachedNioBinaryReader reader = new MultiregionCachedNioBinaryReader(binInputChannel, regions, readCacheSize);
+                        CachedNioBinaryWriter writer = new CachedNioBinaryWriter(tempChannel, writingCacheSize)
+                ) {
+                    doKwayMerge(reader, writer, regions.size());
+                }
+
+            }
+        }
+
+        // now the second pass is simply a single pass with new params
+        try (SeekableByteChannel tempChannel = Files.newByteChannel(tempFile, StandardOpenOption.READ)) {
+            singlePassKwayMerge(
+                    tempChannel,
+                    binOutputChannel,
+                    approxMemLimit,
+                    Math.min(tempChannel.size(), k1 * presortedLen),
+                    writingCacheSize
+            );
+        }
+    }
+
+    // triple pass is basically never practical
```

Comment:
Nope.
Lets base on your buffer of 64K.
For 4GB it will be 64K-way merge.

Two-level tree with arity 64K will fit about 4 * 10^9 elements - less then number of people living on earth.

But 64k is not best possible size of buffer. 
For 512K buffer we get tree with arity 8K and just 64 * 10^6 - less than population of Russia.

And if we consider cheap laptop with 1-2G of avaiable memory ...